loki/transformations/pragma_model.py

```diff
@@ -0,0 +1,324 @@
+        if params_delete := parameters.get('delete'):
+            content += f' map(delete: {params_delete})'
+        if content:
+            return Pragma(keyword='omp', content=f'target exit data{content}')
+        return self.default_retval()
+
+    def pmap_scoped_data(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        content = ''
+        if params_in := parameters.get('in'):
+            content += f' map(to: {params_in})'
+        if params_inout := parameters.get('inout'):
+            content += f' map(tofrom: {params_inout})'
+        if params_out := parameters.get('out'):
+            content += f' map(from: {params_out})'
+        if params_create := parameters.get('create'):
+            content += f' map(alloc: {params_create})'
+        if content:
+            return Pragma(keyword='omp', content=f'target data{content}')
+        return self.default_retval()
+
+    def pmap_end_scoped_data(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        return Pragma(keyword='omp', content='end target data')
+
+    def pmap_routine(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        if 'seq' in parameters:
+            return Pragma(keyword='omp', content='declare target')
+        return self.default_retval()
+
+    def pmap_loop(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        if 'vector' in parameters:
+            # TODO: private and reduction clause?
+            content = 'parallel do'
+            return Pragma(keyword='omp', content=content)
+        if 'gang' in parameters:
+            # TODO: private clause?
+            vlength_param = parameters.get('vlength')
+            vlength = f' thread_limit({vlength_param})' if vlength_param else ''
+            content = f'target teams distribute{vlength}'
+            return Pragma(keyword='omp', content=content)
+        return self.default_retval()
+
+    def pmap_end_loop(self, pragma, parameters, **kwargs): # pylint: disable=unused-argument
+        if 'vector' in parameters:
+            return Pragma(keyword='omp', content='end parallel do')
+        if 'gang' in parameters:
+            return Pragma(keyword='omp', content='end target teams distribute')
+        return self.default_retval()
+
+
+class PragmaModelTransformation(Transformation):
+    """
+    Transformation to map Loki generic pragmas to a specific
+    pragma model using a child class of :any:`GenericPragmaMapper`.
+
+    Parameters
+    __________
+    directive : None, str
+        The directive(s) to be used, used to determine which
+        child class of :any:`GenericPragmaMapper` is used.
+    keep_loki_pragmas: bool
+        Keep or remove generic Loki pragmas that are not
+        mapped.
+    """
+
+    item_filter = (ProcedureItem, ModuleItem)
+
+    def __init__(self, directive=None, keep_loki_pragmas=True, process_module_items=False):
```

Comment:
In what situation would I not want to process module items?

Reply:
I think there is a currently a problem with mode `strict=True` and modules being incomplete. That is why I introduced that.